tooling/bundler/src/bundle/windows/templates/installer.nsi

```diff
@@ -577,6 +577,10 @@ Section Install
   WriteRegDWORD SHCTX "${UNINSTKEY}" "NoRepair" "1"
   WriteRegDWORD SHCTX "${UNINSTKEY}" "EstimatedSize" "${ESTIMATEDSIZE}"
 
+  ; We used to use product name as MAINBINARYNAME
+  ; migrate old shortcuts to target the new MAINBINARYNAME
+  Call MigrateProductNameExeShortcuts
```

Comment:
we don't need to create a separate function for migration, in `CreateStartMenuShortcut` and `CreateDesktopShortcut` just check if it exists and update it or re-create it

Reply:
it doesn't make sense to be in create shortcuts functions, we might skip those functions (we might not currently, but from the name it's not clear it's also responsible for migrating old shortcuts)

Reply:
The name could change, and instead of multiple functions we can just put all logic related to a shortcut inside one function, so we always call it but the logic inside it decides whether to create, update or skip

Reply:
kk

Reply:
Wait, we still need to separate start menu shortcut from desktop one, they have different configs from UIs

Reply:
Yeah 2 functions, one for desktop and one for start menu. I think we need to extract some of the logic into another macro

Reply:
To be honest, I still don't think it make much sense to put migration logic in `CreateStartMenuShortcut`/`CreateDesktopShortcut`, since this is not about creating the shortcut at all (changing those function's name doesn't make much sense to me as well), migration should just be a separate function

Reply:
Problem is to call said functions, we will do the conditional at the call site which sure is okay but bloats the installation section and I would love to hide these in a function so it is easier to focus on just that function

Reply:
Do you mean extract this to a function?

https://github.com/tauri-apps/tauri/blob/040289ea52467c423bd895026d891553409e862a/tooling/bundler/src/bundle/windows/templates/installer.nsi#L582-L595

Reply:
I think I would like it to be
```
   ; Create start menu shortcut 
   !insertmacro MUI_STARTMENU_WRITE_BEGIN Application 
     Call CreateOrUpdateStartMenuShortcut 
   !insertmacro MUI_STARTMENU_WRITE_END 
  
   ; Create desktop shortcut
   Call CreateOrUpdateDesktopShortcut
```

and each of these functions would hide the implementation inside it

Reply:
This wouldn't work, start menu one will be skipped on update

Migration logic here is very different than creating the shortcuts in this context, I suggest we keep shortcut migration code at a central place

Reply:
> This wouldn't work, start menu one will be skipped on update

Why would it be skipped?

Reply:
Doesn't this skip the inner block if no start menu page present (e.g. on update)?

```nsis
   !insertmacro MUI_STARTMENU_WRITE_BEGIN Application 
     Call CreateOrUpdateStartMenuShortcut 
   !insertmacro MUI_STARTMENU_WRITE_END 
```

Reply:
it still calls it, we just skipped page, so think of it as the page was shown for a split second and then clicked next immediately, so the default values are used.

Reply:
You're right, I thought skip if passive means not to run that part, it's actually more or less the same as clicking the next button

But still, it will not be the case anymore once we add in an option to disable that page, I still think put the migration logic together make a lot more sense than putting them in create functions (yes, the name can change, but the caller needs to be more cautious about skipping those function)

Reply:
That's what I am trying to avoid, by always calling them and let the function itself handle the skipping.

Reply:
```nsis
  ; Create start menu shortcut
  !insertmacro MUI_STARTMENU_WRITE_BEGIN Application
    Call CreateStartMenuShortcut
  !insertmacro MUI_STARTMENU_WRITE_END

  ; Create desktop shortcut for silent and passive installers
  ; because finish page will be skipped
  IfSilent create_shortcut 0
  StrCmp $PassiveMode "1" create_shortcut shortcut_done
  create_shortcut:
    Call CreateDesktopShortcut
  shortcut_done:
```

This part itself does some checking and might skip it (I don't want to play with nsis anymore, let's just get it done, in the most clear way)

NSIS itself is so error prone (that's why we use rust to write most things not nsi script), I really don't to rely on those things to work, calling migration here is most clear way of always running it

Reply:
We should always migrate, even the user checks no desktop shortcut at the end, we should still update existing shortcuts

Reply:
I don't agree with that, we shouldn't update if we are not in update mode, otherwise just recreate, reason being if somehow the user messed up their shortcuts and wasn't very experienced to fix it, they will usually resort to reinstalling which should recreate the shortcut and fix it.

Reply:
Current implementation will do exactly that, if user re-runs the installer, it will first migrate the shortcut, and then re-create it, if the user check no update mode, it will only do the migration

Reply:
I know and that is redundant, we could be avoiding that extra work.

Reply:
Alright, let's don't get stuck here debating, I still think it's better to put migration code in a function, but I'll change it to doing it in create shortcut function